server/asset/btc/rpcclient.go

```diff
@@ -130,8 +164,195 @@ func (rc *RPCClient) GetRawTransactionVerbose(txHash *chainhash.Hash) (*btcjson.
+	// case, the most recent block is the best estimate, since fees from older
+	// blocks could become quickly outdated on a busy chain.
+	//
+	// 2) If the first block has fewer transactions, it may be safe to say that
+	// either a) the miner is not stuffing blocks as expected or b) the
+	// blockchain does not have 100 txs/block worth of traffic. Because we have
+	// no historical view of mempool, it's impossible to say which one it is,
+	// though. In this case, taking a weighted average over a few recent blocks
+	// would provide a better estimate.
+
+	var blocksChecked, txCount int
+	var weight uint64
+	for txCount < 101 {
+		if blocksChecked >= rc.maxFeeBlocks {
+			return 0, ErrNoCompetition
+		}
+
+		if err := rc.call(methodGetBlockStats, anylist{blockHash.String(), categories}, &res); err != nil {
+			return 0, err
+		}
+		if len(res.FeeRatePercentiles) != 5 {
+			return 0, fmt.Errorf("unexpected feerate_percentiles response. %d entries", len(res.FeeRatePercentiles))
+		}
+
+		feeRate := res.FeeRatePercentiles[2]
+
+		weight += uint64(res.TxCount) * feeRate
+
+		txCount += res.TxCount
+		blocksChecked++
+
+		if txCount >= 101 {
+			break
+		}
+
+		// Not enough transactions to count yet.
+		verboseBlock, err := rc.GetBlockVerbose(blockHash)
+		if err != nil {
+			return 0, err
+		}
+
+		blockHash, err = chainhash.NewHashFromStr(verboseBlock.PreviousHash)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// rounded average
+	return uint64(math.Round(float64(weight) / float64(txCount))), nil
+}
+
+// medianFeesTheHardWay calculates the median fees from the previous block(s).
+// medianFeesTheHardWay is used for assets that don't have a getblockstats RPC,
+// and is only useful for non-segwit assets.
+func (rc *RPCClient) medianFeesTheHardWay() (uint64, error) {
+	const numTxs = 101
+
+	iHash, err := rc.GetBestBlockHash()
+	if err != nil {
+		return 0, err
+	}
+
+	txs := make([]*wire.MsgTx, 0, numTxs)
+	prevOuts := make(map[chainhash.Hash]map[int]int64, numTxs)
+
+	var blocksChecked int
+
+out:
+	for len(txs) < numTxs {
+		blocksChecked++
+		if blocksChecked > rc.maxFeeBlocks {
+			return 0, ErrNoCompetition
+		}
+
+		blk, err := rc.GetBlockVerbose(iHash)
+		if err != nil {
+			return 0, err
+		}
+
+		if len(blk.Tx) == 0 {
+			return 0, fmt.Errorf("no transactions?")
+		}
+
+		rawTxs := blk.Tx[1:] // skip coinbase
+		rand.Shuffle(len(rawTxs), func(i, j int) { rawTxs[i], rawTxs[j] = rawTxs[j], rawTxs[i] })
+
+		for _, txid := range rawTxs {
+			txHash, err := chainhash.NewHashFromStr(txid)
+			if err != nil {
+				return 0, err
+			}
+
+			utilTx, err := rc.GetRawTransaction(txHash)
+			if err != nil {
+				return 0, err
+			}
+
+			tx := utilTx.MsgTx()
+
+			for _, vin := range tx.TxIn {
+				prevOut := vin.PreviousOutPoint
+				prevs := prevOuts[prevOut.Hash]
+				if len(prevs) == 0 {
+					prevs = make(map[int]int64, 1)
+					prevOuts[prevOut.Hash] = prevs
+				}
+				prevs[int(prevOut.Index)] = 0 // placeholder
```

Comment:
`// value will be retrieved subsequently, after all prevouts for this funding txhash are recorded`